util/clusterauth/clusterauth.go

```diff
@@ -212,42 +216,119 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
-	var serviceAccount *corev1.ServiceAccount
+// GetServiceAccountBearerToken determines if a ServiceAccount has a
+// bearer token secret to use or if a secret should be created. It then
+// waits for the secret to have a bearer token if a secret needs to
+// be created and returns the token in encoded base64.
+func GetServiceAccountBearerToken(clientset kubernetes.Interface, ns string, sa string, timeout time.Duration) (string, error) {
+	secretName, err := getOrCreateServiceAccountTokenSecret(clientset, sa, ns)
+	if err != nil {
+		return "", err
+	}
+
 	var secret *corev1.Secret
-	var err error
-	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
-		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+	err = wait.PollImmediate(500*time.Millisecond, timeout, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), ClientGoRequestTimeout)
+		defer cancel()
+		secret, err = clientset.CoreV1().Secrets(ns).Get(ctx, secretName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to get secret %q for serviceaccount %q: %w", secretName, sa, err)
 		}
-		// Scan all secrets looking for one of the correct type:
-		for _, oRef := range serviceAccount.Secrets {
-			var getErr error
-			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
-			if err != nil {
-				return false, fmt.Errorf("Failed to retrieve secret %q: %v", oRef.Name, getErr)
-			}
-			if secret.Type == corev1.SecretTypeServiceAccountToken {
-				return true, nil
-			}
+
+		_, ok := secret.Data["token"]
+		if !ok {
+			return false, nil
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get token for serviceaccount %q: %w", sa, err)
+	}
+
+	return string(secret.Data["token"]), nil
+}
+
+// getOrCreateServiceAccountTokenSecret will check if a ServiceAccount
+// already has a kubernetes.io/service-account-token secret associated
+// with it or creates one if the ServiceAccount doesn't have one. This
+// was added to help add k8s v1.24+ clusters.
+func getOrCreateServiceAccountTokenSecret(clientset kubernetes.Interface, sa, ns string) (string, error) {
+	// Wait for sa to have secret, but don't wait too
+	// long for 1.24+ clusters
+	var serviceAccount *corev1.ServiceAccount
+	err := wait.PollImmediate(500*time.Millisecond, 5*time.Second, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), ClientGoRequestTimeout)
+		defer cancel()
+		var getErr error
+		serviceAccount, getErr = clientset.CoreV1().ServiceAccounts(ns).Get(ctx, sa, metav1.GetOptions{})
+		if getErr != nil {
+			return false, fmt.Errorf("failed to get serviceaccount %q: %w", sa, getErr)
 		}
-		return false, nil
+		if len(serviceAccount.Secrets) == 0 {
+			return false, nil
+		}
+		return true, nil
 	})
+	if serviceAccount == nil {
```

Comment:
TODO: Move err check on line 277 out of this loop.

Reply:
@leoluz I think we still need to keep the current approach here in order to prevent the timeout from wait.PollImmediate(). Adding if err != nil leads to tests failing from the timeout. Let me know if I misunderstood.

Reply:
Strange. 
What if the code is like:
```go
if err != nil {
	return "", fmt.Errorf("failed to create serviceaccount token secret: %w", err)
}
if serviceAccount == nil {
    log.Errorf("Unexpected nil serviceaccount '%s/%s' with no error returned", ns, sa)
    return "", fmt.Errorf("failed to create serviceaccount token secret: nil serviceaccount returned for '%s/%s' with no error", ns, sa)
}
```

Reply:
I think the problem is just the timeout from wait.PollImmediate will prevent this from moving on to the secret creation if err != nil comes immediately after.

Reply:
@danielhelfand you should be able to check the timeout error by the returned error type:
https://github.com/kubernetes/apimachinery/blob/v0.24.1/pkg/util/wait/wait.go#L205:5
It would be something like:
```golang
if err != nil && err != wait.ErrWaitTimeout {
	return "", fmt.Errorf("failed to create serviceaccount token secret: %w", err)
}
if serviceAccount == nil {
    log.Errorf("Unexpected nil serviceaccount '%s/%s' with no error returned", ns, sa)
    return "", fmt.Errorf("failed to create serviceaccount token secret: nil serviceaccount returned for '%s/%s' with no error", ns, sa)
}
```

Reply:
Thanks! I have gone ahead and added.